libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteProgress.kt

```diff
@@ -12,23 +12,23 @@ import com.mapbox.geojson.Point
  * With every new valid location update, a new route progress will be generated using the latest
  * information.
  *
- * The latest route progress object can be obtained through the [com.mapbox.navigation.base.trip.RouteProgressObserver].
+ * The latest route progress object can be obtained through the [com.mapbox.navigation.core.trip.session.RouteProgressObserver].
  * Note that the route progress object's immutable.
  */
-class RouteProgress private constructor(
-    private val route: DirectionsRoute? = null,
-    private val routeGeometryWithBuffer: Geometry? = null,
-    private val bannerInstructions: BannerInstructions? = null,
-    private val voiceInstructions: VoiceInstructions? = null,
-    private val currentState: RouteProgressState? = null,
-    private val currentLegProgress: RouteLegProgress? = null,
-    private val upcomingStepPoints: List<Point>? = null,
-    private val inTunnel: Boolean = false,
-    private val distanceRemaining: Float = 0f,
-    private val distanceTraveled: Float = 0f,
-    private val durationRemaining: Double = 0.0,
-    private val fractionTraveled: Float = 0f,
-    private val remainingWaypoints: Int = 0,
+data class RouteProgress(
```

Comment:
Should this be `private` to enforce the use of the `Builder`?

Reply:
Yeah I was also tried this. With data classes, private is meaningless because of `copy()`

![Screen Shot 2020-05-18 at 4 43 37 PM](https://user-images.githubusercontent.com/3021882/82269491-bb653300-9926-11ea-90a7-a394ab59604c.png)

In fact, the `toBuilder` might be deceiving with this pattern because because of `copy()`. Once you have the data object, you can create a new one.
``` kotlin
val newRouteProgress = routeProgress.copy(
  distanceRemaining = 145.0f
)
```

Reply:
Yeah the point is what is preventing you to create a `RouteProgress` directly using the primary constructor instead of the `Builder`. I think that was the main reason for being a `class` instead of a `data class` 👀 https://github.com/mapbox/mapbox-navigation-android/pull/2624#discussion_r403145619 cc @cafesilencio

Reply:
> Yeah the point is what is preventing you to create a RouteProgress directly using the primary constructor instead of the Builder. 

Do we want to prevent external customers from being able to create `RouteProgress` objects? This is a fair question because we may have two categories of public domain entities.

I've added a `Goal` section to the PR description. Because it seems we may have identified domain entities with different requirements. `RouteProgress`, `RouteStepProgress`, .. these objects have a different responsibility from parameter objects like `NavigationOptions` and `ArrivalOptions`.

Reply:
> Do we want to prevent external customers from being able to create `RouteProgress` objects?

I'd say so, right? These are created internally and should be read-only for clients. That being said, we can always let developers to create them without the need of going through the builder, as cannot be injected. Is that what you were thinking?

Reply:
🤔 i can see the separation between the two types of public objects. Looking for an example where we would want to make the limitation for the objects generated by navigation.

All of the directions api responses can be rebuilt, using the Builder may be something that is needed for java. But I'm not seeing why we would want to prevent people from using `copy()` yet

Reply:
Don't get me wrong my point was around having a `private constructor` and forcing the `Builder` usage, if pros of having `copy` beat cons, let's do it :+1:

Reply:
> All of the directions api responses can be rebuilt, using the Builder may be something that is needed for java.

Yeah and you cannot create the object directly.

Also :eyes: [Public API challenges in Kotlin]( https://jakewharton.com/public-api-challenges-in-kotlin/)